File: pwa/src/lib/notifications.ts
import { getMetadata, setMetadata } from '$lib/db/metadata';
import type { TaskRecord } from '$lib/db/schema';

const META_DAILY = 'last_daily_notify';
const META_NOTIFIED = 'notified_tasks';

/** Request notification permission. Call once on mount. */
export async function requestNotificationPermission(): Promise<boolean> {
  if (!('Notification' in window)) return false;
  if (Notification.permission === 'granted') return true;
  if (Notification.permission === 'denied') return false;
  const result = await Notification.requestPermission();
  return result === 'granted';
}

/**
 * Check all three notification moments after a sync completes.
 * Must be called with the current list of pending tasks.
 */
export async function checkNotifications(pendingTasks: TaskRecord[]): Promise<void> {
  if (!('Notification' in window) || Notification.permission !== 'granted') return;

  const now = new Date();
  const today = localDate(now);

  // ── 1. Morning Brief (first sync of the day) ──
  const lastBrief = await getMetadata(META_DAILY);
  if (today !== lastBrief) {
    const todayTasks = pendingTasks.filter(
      (t) => t.due_at && localDate(new Date(t.due_at)) === today
    );
    if (todayTasks.length > 0) {
      const titles = todayTasks.slice(0, 2).map((t) => t.title);
      const rest = todayTasks.length - titles.length;
      let text = 'Today: ' + titles.join(', ');
      if (rest === 1) text += ' and 1 more';
      else if (rest > 1) text += ` and ${rest} more`;
      sendNotification(text);
    }
    await setMetadata(META_DAILY, today);
  }

  // ── 2 & 3. Task due / overdue ──
  // notified_tasks maps task UUID → due_at ISO string (so we can detect reschedules)
  const raw = await getMetadata(META_NOTIFIED);
  const notified: Record<string, string> = raw ? JSON.parse(raw) : {};
  const cleaned: Record<string, string> = {};

  for (const task of pendingTasks) {
    const taskId = task.id;
    const storedDue = notified[taskId];

    if (storedDue !== undefined) {
      // Task was previously notified. Check if the due date has changed.
      // If it changed, the old notification is stale – drop it.
      // If it's the same, keep the entry to prevent re-notification.
      if (task.due_at === storedDue) {
        cleaned[taskId] = storedDue;
      }
      // If due_at changed, don't carry over the entry. The new due date
      // will trigger a fresh notification when it arrives.
      continue;
    }

    if (!task.due_at) continue;

    const dueAt = new Date(task.due_at);
    const hasTime = dueAt.getUTCHours() !== 0 || dueAt.getUTCMinutes() !== 0;

    if (hasTime) {
      // Task has a specific time — check the 5-minute "due" window
      const fiveMinAfter = new Date(dueAt.getTime() + 5 * 60 * 1000);
      if (now >= dueAt && now < fiveMinAfter) {
        // Moment 2: Task becomes due
        sendNotification('Time: ' + task.title);
        cleaned[taskId] = task.due_at;
      } else if (now >= fiveMinAfter) {
        // Moment 3: Task overdue (missed the due window)
        sendNotification(task.title + ' is overdue');
        cleaned[taskId] = task.due_at;
      }
    } else {
      // No specific time (midnight) — check if the date has passed
      if (localDate(dueAt) < today) {
        sendNotification(task.title + ' is overdue');
        cleaned[taskId] = task.due_at;
      }
      // Tasks due today without a time are covered by the morning brief
    }
  }

  await setMetadata(META_NOTIFIED, JSON.stringify(cleaned));
}

function sendNotification(body: string): void {
  try {
    new Notification('TodoRS', {
      body,
      icon: '/icon-192.png',
      badge: '/icon-192.png',
      tag: 'todors-' + Date.now(),
    });
  } catch {
    // Silently ignore if notification fails (e.g. permission revoked mid-session)
  }
}

/** Return the local date as YYYY-MM-DD. */
function localDate(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

File: crates/todomrs-tui/src/notifications.rs
use anyhow::Result;
use chrono::{Local, Timelike, Utc};
use notify_rust::Notification;
use sqlx::SqlitePool;
use std::collections::HashMap;
use todomrs_core::domain::Task;

/// Check all three notification moments after a sync completes.
///
/// Uses the metadata table in SQLite for state tracking, same keys as the PWA.
///
/// # Per-account safety
///
/// The metadata keys `last_daily_notify` and `notified_tasks` are read from
/// and written to the per-account database. Since each authenticated account
/// opens a different SQLite file, notification state cannot bleed across
/// accounts. This was previously a cross-account concern when all accounts
/// shared one global database.
pub async fn check_notifications(pool: &SqlitePool, tasks: &[Task]) -> Result<()> {
    let pending: Vec<&Task> = tasks
        .iter()
        .filter(|t| t.status == todomrs_core::domain::TaskStatus::Pending && t.deleted_at.is_none())
        .collect();

    let now = Utc::now();
    let today = local_date_today();

    // ── 1. Morning Brief ──
    let last_brief = get_metadata(pool, "last_daily_notify").await?;
    if last_brief.as_deref() != Some(&today) {
        let today_tasks: Vec<&&Task> = pending
            .iter()
            .filter(|t| {
                t.due_at
                    .map(|d| format_local_date(d))
                    .as_deref()
                    == Some(&today)
            })
            .collect();

        if !today_tasks.is_empty() {
            let titles: Vec<&str> = today_tasks.iter().take(2).map(|t| t.title.as_str()).collect();
            let rest = today_tasks.len() - titles.len();
            let mut text = format!("Today: {}", titles.join(", "));
            if rest == 1 {
                text.push_str(" and 1 more");
            } else if rest > 1 {
                text.push_str(&format!(" and {} more", rest));
            }
            let _ = send_notification(&text);
        }
        set_metadata(pool, "last_daily_notify", &today).await?;
    }

    // ── 2 & 3. Task due / overdue ──
    let notified_json = get_metadata(pool, "notified_tasks").await?;
    let notified: HashMap<String, String> = notified_json
        .as_deref()
        .and_then(|s| serde_json::from_str(s).ok())
        .unwrap_or_default();
    let mut cleaned: HashMap<String, String> = HashMap::new();

    for task in &pending {
        let task_id = task.id.to_string();
        let stored_due = notified.get(&task_id);

        if let Some(stored) = stored_due {
            // Previously notified. Keep only if the due_at hasn't changed.
            let current_due = task.due_at.map(|d| d.to_rfc3339());
            if current_due.as_deref() == Some(stored) {
                cleaned.insert(task_id, stored.clone());
            }
            // If due_at changed, drop the entry. New due date = fresh notification.
            continue;
        }

        let due_at = match task.due_at {
            Some(d) => d,
            None => continue,
        };

        let has_time = due_at.hour() != 0 || due_at.minute() != 0;

        if has_time {
            let five_min_later = due_at + chrono::Duration::minutes(5);
            if now >= due_at && now < five_min_later {
                // Moment 2: Task becomes due
                let _ = send_notification(&format!("Time: {}", task.title));
                cleaned.insert(task_id, due_at.to_rfc3339());
            } else if now >= five_min_later {
                // Moment 3: Task overdue
                let _ = send_notification(&format!("{} is overdue", task.title));
                cleaned.insert(task_id, due_at.to_rfc3339());
            }
        } else {
            // No time component — midnight
            let due_date = format_local_date(due_at);
            if due_date < today {
                // Moment 3: Date passed
                let _ = send_notification(&format!("{} is overdue", task.title));
                cleaned.insert(task_id, due_at.to_rfc3339());
            }
            // Tasks due today without a time: covered by morning brief
        }
    }

    let cleaned_json = serde_json::to_string(&cleaned)?;
    set_metadata(pool, "notified_tasks", &cleaned_json).await?;

    Ok(())
}

fn send_notification(body: &str) -> Result<()> {
    Notification::new()
        .summary("TodoRS")
        .body(body)
        .show()?;
    Ok(())
}

/// Return today's local date as YYYY-MM-DD.
fn local_date_today() -> String {
    Local::now().format("%Y-%m-%d").to_string()
}

/// Format a datetime's local date as YYYY-MM-DD.
fn format_local_date(dt: chrono::DateTime<Utc>) -> String {
    dt.with_timezone(&Local).format("%Y-%m-%d").to_string()
}

// ── Metadata helpers (reuse the existing metadata table) ──

async fn get_metadata(pool: &SqlitePool, key: &str) -> Result<Option<String>> {
    let row: Option<(String,)> = sqlx::query_as("SELECT value FROM metadata WHERE key = ?")
        .bind(key)
        .fetch_optional(pool)
        .await?;
    Ok(row.map(|r| r.0))
}

async fn set_metadata(pool: &SqlitePool, key: &str, value: &str) -> Result<()> {
    sqlx::query("INSERT OR REPLACE INTO metadata (key, value) VALUES (?, ?)")
        .bind(key)
        .bind(value)
        .execute(pool)
        .await?;
    Ok(())
}
